velox/exec/Driver.cpp

```diff
@@ -348,25 +356,42 @@ StopReason Driver::runInternal(
             uint64_t resultBytes = 0;
             RowVectorPtr result;
             {
-              auto timer = cpuWallTimer(op->stats().getOutputTiming);
+              auto timer = cpuWallTimerCopy(
+                  op->stats().rlock()->getOutputTiming,
+                  [op](const CpuWallTiming& timing) {
+                    op->stats().wlock()->getOutputTiming = timing;
+                  });
+              runtimeStatWriter->setActiveOperator(op);
               result = op->getOutput();
               if (result) {
                 VELOX_CHECK(
                     result->size() > 0,
-                    "Operator::getOutput() must return nullptr or a non-empty vector: {}",
-                    op->stats().operatorType);
-                op->stats().outputVectors += 1;
-                op->stats().outputPositions += result->size();
+                    "Operator::getOutput() must return nullptr or "
+                    "a non-empty vector: {}",
+                    op->operatorType());
                 resultBytes = result->estimateFlatSize();
-                op->stats().outputBytes += resultBytes;
+                {
+                  auto lockedStats = op->stats().wlock();
+                  lockedStats->outputVectors += 1;
+                  lockedStats->outputPositions += result->size();
+                  lockedStats->outputBytes += resultBytes;
+                }
               }
             }
             pushdownFilters(i);
             if (result) {
-              auto timer = cpuWallTimer(op->stats().addInputTiming);
-              nextOp->stats().inputVectors += 1;
-              nextOp->stats().inputPositions += result->size();
-              nextOp->stats().inputBytes += resultBytes;
+              auto timer = cpuWallTimerCopy(
+                  nextOp->stats().rlock()->addInputTiming,
+                  [nextOp](const CpuWallTiming& timing) {
+                    nextOp->stats().wlock()->addInputTiming = timing;
+                  });
+              {
+                auto lockedStats = nextOp->stats().wlock();
+                lockedStats->inputVectors += 1;
+                lockedStats->inputPositions += result->size();
+                lockedStats->inputBytes += resultBytes;
+              }
+              runtimeStatWriter->setActiveOperator(nextOp);
```

Comment:
```
How about we use scoped data structure to achieve this and then we don't need setActiveOperator method and statWriterGuard? Thanks!

class ScopedOpRuntimeStatWriter {
  ScopedOpRuntimeStatWriter(Operator* op) {
     setThreadLocalRunTimeStatWriter(op);
  }
  ~ScopedOpRuntimeStatWriter() {
    setThreadLocalRunTimeStatWriter(nullptr);
  }
}

#define ADD_OP_INPUT(op, input)  \
 do {
    ScopedOpRuntimeStatWriter runtimeStatWriter(op);
    op->addInput(input);
 } while (0);

#define GET_OP_OUTPUT(op, result, timeStats) \
do {
    ScopedOpRuntimeStatWriter runtimeStatWriter(op);
    result = op->getOutput();
} while (0);

Similar for the other op method invocations.
```

Reply:
I would rather not complicate this with the scope thingy now.
What would we set the operator to when we leave the scope? A previous one? A null? Not sure.
Also the using of do/while loop is weird?
Also what is op->addInput(input); and result = op->getOutput();?
Are you suggesting on pairing these calls?
It won't work well, because we need to set active operator before ALL calls to an operator, since we can have counters added pretty much in any member (not just addInput() and getOutput() and we actually have this already. :)